terraform/gcp/buckets.tf

```diff
@@ -38,13 +38,32 @@ locals {
   ]))
 }
 
+locals {
+  # Get mapping of public buckets and their associated hub names
+  public_buckets = distinct(flatten([
+    for hub_name, permissions in var.hub_cloud_permissions : [
+      for bucket_name in permissions.public_buckets : {
+        hub_name    = hub_name
+        bucket_name = bucket_name
+      }
+    ]
+  ]))
+}
+
 resource "google_storage_bucket_iam_member" "member" {
   for_each = { for bp in local.bucket_permissions : "${bp.hub_name}.${bp.bucket_name}" => bp }
   bucket   = google_storage_bucket.user_buckets[each.value.bucket_name].name
   role     = "roles/storage.admin"
   member   = "serviceAccount:${google_service_account.workload_sa[each.value.hub_name].email}"
 }
 
+resource "google_storage_default_object_access_control" "public_rule" {
+  for_each = { for bp in local.public_buckets : "${bp.hub_name}.${bp.bucket_name}" => bp }
```

Comment:
I used the same mapping mechanism to get pairs of [hub, public buckets], but I don't think it makes sense to have this in the context of a public bucket.
Or is there a way to enforce a namespace on a scratch bucket?

Reply:
I'm new to terraform code, so I'm for actively learning as part of this review.

### References

- Read that [`for_each`](https://developer.hashicorp.com/terraform/language/meta-arguments/for_each) has a special meaning, available for all resources
- I understand that [`bucket` / `role` / `member`](https://registry.terraform.io/providers/hashicorp/google/latest/docs/resources/storage_bucket_iam#google_storage_bucket_iam_member) has resource type (`google_storage_default_object_access_control`) specific meaning.

### Refactoring to avoid complexity

Looking at this resource definition and the one above, I see that `for_each` is used to provide a key and value (key => value), but we never make use of the key (each.key). So, we could switch this to a statement like this instead. Both here and above. That would have made me understand this section faster, and probably myself in the future as well.

```diff
-  for_each = { for bp in local.public_buckets : "${bp.hub_name}.${bp.bucket_name}" => bp }
+  for_each = [ for b in local.public_buckets : b ]
   bucket = google_storage_bucket.user_buckets[each.value.bucket_name].name
   role   = "READER"
   entity = "allUsers"
 }
```

I think actually this may work also.

```diff
-  for_each = { for bp in local.public_buckets : "${bp.hub_name}.${bp.bucket_name}" => bp }
+  for_each = local.public_buckets
   bucket = google_storage_bucket.user_buckets[each.value.bucket_name].name
   role   = "READER"
   entity = "allUsers"
 }
```

- [ ] Refactor for readability like above, do it also for the other resource for consistency.

### Namespacing

It doesn't make sense for the `google_storage_default_object_access_control` resource to consider namespace. The purpose is to declare public readability.

On the other hand, we still need to provide hub users access to write to the bucket which is declared in the `google_storage_bucket_iam_member` above. Looking at that, I conclude that permission is only granted to a namespaced account for buckets listed in `bucket_admin_access` - where the public bucket isn't listed!

So, the public bucket still needs to be provisioned write access via namespaced IAM accounts (coupled to a k8s ServiceAccount via google workload-identity).

- [ ] Add the public bucket name to the `bucket_admin_access` list for m2lines